File: src/coreclr/vm/cdacmetadata.cpp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

#include "cdacmetadata.hpp"

#ifndef DACCESS_COMPILE
CDacMetadata g_cdacMetadata;

void CDacMetadata::Init()
{
    PrecodeMachineDescriptor::Init(&g_cdacMetadata.precode);
#if defined(TARGET_ARM)
    g_cdacMetadata.codePointerFlags = CDacCodePointerFlags::HasArm32ThumbBit;
#elif defined(TARGET_ARM64) && defined(TARGET_OSX)
    g_cdacMetadata.codePointerFlags = CDacCodePointerFlags::None; // TODO set HasArm64PtrAuth if arm64e
#else
    g_cdacMetadata.codePointerFlags = CDacCodePointerFlags::None;
#endif
}

#endif // !DACCESS_COMPILE

File: src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Data/CDacMetadata.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

namespace Microsoft.Diagnostics.DataContractReader.Data;

internal sealed class CDacMetadata : IData<CDacMetadata>
{
    static CDacMetadata IData<CDacMetadata>.Create(Target target, TargetPointer address)
        => new CDacMetadata(target, address);

    public CDacMetadata(Target target, TargetPointer address)
    {
        Target.TypeInfo type = target.GetTypeInfo(DataType.CDacMetadata);
        PrecodeMachineDescriptor = address + (ulong)type.Fields[nameof(PrecodeMachineDescriptor)].Offset;
        CodePointerFlags = target.Read<byte>(address + (ulong)type.Fields[nameof(CodePointerFlags)].Offset);
    }

    /* Address of */
    public TargetPointer PrecodeMachineDescriptor { get; init; }
    public byte CodePointerFlags { get; init; }
}
